File: Platform/codesignal/Intro/21.isIPv4Address/solution.js
function solution(s) {
  arr = s.split('.')
  if (arr.length !== 4) return false
  return arr.map(x => {
    if (isNaN(x)) return false;
    if (x === "") return false;
    if (x < 0 || x > 255) return false;
    if (x.length !== LenNum(x)) return false;
    return true
  }).every(x=>x)
}

function LenNum(x) {
  let result = 1;
  while(x >= 10) {
    x = ~~(x/10)
    result++
  }
  return result
}
// console.log(solution("172.16.254.14"));
// console.log(solution(".254.255.0"));
console.log(solution("00.1.1.1a"));

File: Platform/codesignal/Intro/21.isIPv4Address/solution.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func solution(s string) bool {
	arr := strings.Split(s, ".")
	if len(arr) != 4 {
		return false
	}
	for i := 0; i < len(arr); i++ {
		if arr[i] == "" {
			return false
		}
		num, err := strconv.Atoi(arr[i])
		if err != nil {
			return false
		}
		if num < 0 || num > 255 {
			return false
		}
		if len(arr[i]) != LenNum(num) {
			return false
		}
	}
	return true
}

func LenNum(s int) int {
	result := 1
	for s >= 10 {
		s = s / 10
		result++
	}
	return result
}

func main() {
	str := "172.16.254.1"
	fmt.Println(solution(str))
}
